common/src/main/java/io/netty/util/internal/PlatformDependent0.java

```diff
@@ -415,7 +437,605 @@ static void freeMemory(long address) {
+                    } catch (NoSuchFieldException e) {
+                        logger.warn("Failed to find String value array." +
+                                "String hash code optimizations are disabled.", e);
+                    } catch (SecurityException e) {
+                        logger.info("No permissions to get String value array." +
+                                "String hash code optimizations are disabled.", e);
+                    }
+                    return null;
+                }
+            });
+            Field stringBuilderValueField = AccessController.doPrivileged(new PrivilegedAction<Field>() {
+                @Override
+                public Field run() {
+                    try {
+                        Field f = StringBuilder.class.getSuperclass().getDeclaredField("value");
+                        f.setAccessible(true);
+                        return f;
+                    } catch (NoSuchFieldException e) {
+                        logger.warn("Failed to find StringBuilder value array." +
+                                " Hash code optimizations for this type are disabled.", e);
+                    } catch (SecurityException e) {
+                        logger.info("No permissions to get StringBuilder value array." +
+                                " Hash code optimizations for this type are disabled.", e);
+                    }
+                    return null;
+                }
+            });
+            STRING_VALUE_FIELD_OFFSET = stringValueField == null ?
+                    -1 : PlatformDependent0.objectFieldOffset(stringValueField);
+            STRING_BUILDER_VALUE_FIELD_OFFSET = stringBuilderValueField == null ?
+                    -1 : PlatformDependent0.objectFieldOffset(stringBuilderValueField);
+        }
+
+        @Override
+        public final int emptyHashValue() {
+            return EMPTY_HASH;
+        }
+
+        @Override
+        public final int hashCodeAsBytes(CharSequence data, int startPos, int endPos) {
+            // CharSequence by itself does not provide access to any underlying data structure, and using
+            // charAt can be costly. For the common types in the CharSequence we can provide an optimized
+            // hashCode by direclty using the underlying char[].
+            if (data.getClass() == String.class && STRING_VALUE_FIELD_OFFSET != -1) {
+                return hashCodeAsBytes(
+                        (char[]) UNSAFE.getObject(data, STRING_VALUE_FIELD_OFFSET), startPos, endPos);
+            }
+            if (data.getClass() == StringBuilder.class && STRING_BUILDER_VALUE_FIELD_OFFSET != -1) {
+                return hashCodeAsBytes(
+                        (char[]) UNSAFE.getObject(data, STRING_BUILDER_VALUE_FIELD_OFFSET), startPos, endPos);
+            }
+            return hashCodeAsBytesDirect(data, startPos, endPos);
+        }
+
+        protected abstract int hashCodeAsBytesDirect(CharSequence data, int startPos, int endPos);
```

Comment:
Hah sounds about right. I think case it may be [blameScott](https://github.com/netty/netty/pull/3626#issuecomment-93861594) or [blameTrustin](https://github.com/netty/netty/pull/3626#issuecomment-93864908) though :)